File: Python/Strings/L_determice2StrSame.py
def closeStrings(word1: str, word2: str) -> bool:
        if len(word1) != len(word2):
            return False

        cnt1,cnt2 = {}, {}
        for ch in word1:
            cnt1[ch] = 1 + cnt1.get(ch, 0)
        for ch in word2:
            cnt2[ch] = 1 + cnt2.get(ch, 0)       
            
        res1 = [val for val in cnt1.values()]
        res2 = [val for val in cnt2.values()]
        res1.sort()
        res2.sort()
        return res1 == res2 and set(word1) == set(word2)
    
    
def closeStrings(word1: str, word2: str) -> bool:
        freq1 = [0] * 26
        freq2 = [0] * 26

        for ch in word1:
            freq1[ord(ch) - ord('a')] += 1

        for ch in word2:
            freq2[ord(ch) - ord('a')] += 1

        for i in range(26):
            if (freq1[i] == 0 and freq2[i] != 0) or (freq1[i] != 0 and freq2[i] == 0):
                return False

        freq1.sort()
        freq2.sort()

        for i in range(26):
            if freq1[i] != freq2[i]:
                return False

        return True

File: cpp/String/L_determice2StrSame.cpp
#include <bits/stdc++.h>
using namespace std;

bool closeStrings(std::string word1, std::string word2)
{
    std::vector<int> freq1(26, 0);
    std::vector<int> freq2(26, 0);

    for (char ch : word1)
    {
        freq1[ch - 'a']++;
    }

    for (char ch : word2)
    {
        freq2[ch - 'a']++;
    }

    for (int i = 0; i < 26; i++)
    {
        if ((freq1[i] == 0 && freq2[i] != 0) || (freq1[i] != 0 && freq2[i] == 0))
        {
            return false;
        }
    }

    std::sort(freq1.begin(), freq1.end());
    std::sort(freq2.begin(), freq2.end());

    for (int i = 0; i < 26; i++)
    {
        if (freq1[i] != freq2[i])
        {
            return false;
        }
    }

    return true;
}

bool closeStrings1(string word1, string word2)
{
    unordered_map<char, int> mp1, mp2, mp3, mp4;
    // existence checking
    for (auto i : word1)
        mp1[i]++, mp3[i] = 1;
    for (auto i : word2)
        mp2[i]++, mp4[i] = 1;
    if (mp4 != mp3)
        return false;
    // checking if we can swap
    multiset<int> s1, s2;
    for (auto [c, f] : mp1)
        s1.insert(f);
    for (auto [c, f] : mp2)
        s2.insert(f);
    return s1 == s2;
}